src/main/drivers/rssi_softpwm.c

```diff
@@ -0,0 +1,272 @@
+        } else if (rawWidth < pulseMin) {
+            pulseWidth = pulseMin;
+        } else {
+            pulseWidth = rawWidth;
+        }
+        break;
+
+    case RXTYPE_FRSKY_X4R: // Probably all X series
+
+        // X4R's RSSI signal is evil; shortest pulse is less than 3usec,
+        // which is beyond the limit of software assisted PWM input can
+        // reliably measure (probably even with input capture).
+        // Here, we admit that we can miss the falling edge of the pulse
+        // we have started our measurement, and compensate for the extra
+        // cycle (effectively measuring the next pulse provided cycle time
+        // is stable).
+
+        if (rawWidth >= 990) {
+            // Must have missed the first falling edge
+            if (rawWidth < 1000 + 100) {
+                if (rawWidth < 1000)
+                    pulseWidth = 0;
+                else
+                    pulseWidth = rawWidth - 1000;
+            } else {
+                // Something went wrong
+                pulseWidth = 0; // Fail
+            }
+        } else {
+            pulseWidth = rawWidth;
+        }
+        break;
+
+    default:
+        pulseWidth = 500; // Default marker
+        break;
+    }
+
+    debug[1] = pulseWidth;
+
+    // Monitor pulseWidths for parameter decision.
+
+    if (rawWidth > tmax)
+        tmax = rawWidth;
+
+    if (rawWidth < tmin)
+        tmin = rawWidth;
+
+    debug[4] = tmax;
+}
+
+static int32_t rspFilterArray[5];
+static int rspFilterPos;
+
+bool rssiSoftPwmInit(void)
+{
+    // Nothing special, standard spell.
+    // XXX Are these done in exti.c already?
+#ifdef STM32F10X
+    // enable AFIO for EXTI support
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_AFIO, ENABLE);
+#endif
+
+#if defined(STM32F3) || defined(STM32F4)
+    /* Enable SYSCFG clock otherwise the EXTI irq handlers are not called */
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SYSCFG, ENABLE);
+#endif
+
+    ioTag_t tag = masterConfig.rssiSoftPwmConfig.ioTag;
+    if (tag == IO_TAG_NONE)
+        return false;
+
+    rspIO = IOGetByTag(tag);
+
+    IOInit(rspIO, OWNER_RSSIPWM, 0);
+    IOConfigGPIO(rspIO, IOCFG_IN_FLOATING);
+
+    EXTIHandlerInit(&rsp_extiCallbackRec, rspExtiHandler);
+    EXTIConfig(rspIO, &rsp_extiCallbackRec, NVIC_PRIO_SOFTPWM_EXTI, EXTI_Trigger_Rising);
+
+    rxtype = masterConfig.rssiSoftPwmConfig.device;
+
+    switch(rxtype) {
+    case RXTYPE_FRSKY_TFR4:
+        pulseMin = 24;
+        pulseMax = 1024;
+        break;
+
+    default:
+        pulseMin = 1;
+        pulseMax = 999;
+        break;
+    }
+
+    rspInitialized = true;
+    rspActive = false;
+
+    for (int i = 0 ; i < 5 ; i++)
+        rspFilterArray[i] = 0;
+
+    rspFilterPos = 0;
+
+    return true;
+}
+
+// rssiSoftPwmUpdate: Scheduler task
+
+void rssiSoftPwmUpdate(uint32_t currentTime)
+{
+    UNUSED(currentTime);
+
+    uint32_t value;
+    bool longPulse = false;
+    uint8_t longValue;
+
+    if (!rspInitialized) {
```

Comment:
IMO it should be impossible to call task function without calling Init first. `rspInitialized` only complicating things ...